src/derive_builder.rs

```diff
@@ -0,0 +1,158 @@
+/// Derive a builder for a struct
+///
+/// Use in combiantion with [custom_derive][custom_derive].
+///
+/// [custom_derive]: https://crates.io/crates/custom_derive
+///
+/// # Examples
+///
+/// ```rust
+/// #[macro_use] extern crate custom_derive;
+/// #[macro_use] extern crate derive_builder;
+///
+/// custom_derive!{
+///     #[derive(Debug, PartialEq, Default, Builder)]
+///     struct Lorem {
+///         ipsum: String,
+///         dolor: i32,
+///     }
+/// }
+///
+/// fn main() {
+///     let x = Lorem::default().ipsum("sit").dolor(42);
+///     assert_eq!(x, Lorem { ipsum: "sit".into(), dolor: 42 });
+/// }
+/// ```
+///
+/// ## Generic structs
+///
+/// ```rust
+/// #[macro_use] extern crate custom_derive;
+/// #[macro_use] extern crate derive_builder;
+///
+/// custom_derive!{
+///     #[derive(Debug, PartialEq, Default, Builder)]
+///     struct GenLorem<T> {
+///         ipsum: String,
+///         dolor: T,
+///     }
+/// }
+///
+/// fn main() {
+///     let x = GenLorem::default().ipsum("sit").dolor(42);
+///     assert_eq!(x, GenLorem { ipsum: "sit".into(), dolor: 42 });
+/// }
+/// ```
+///
+/// ## Gotchas
+///
+/// - Tuple structs and unit structs are not supported as they have no field
+///   names.
+/// - When defining a generic struct, you cannot use `V` as a generic parameter
+///   as this is what all setters are using.
```

Comment:
I think `impl<T>` means `T` is declared for the entire `impl` and all its `fns`. [This code](https://play.rust-lang.org/?code=struct%20Foo%3CT%3E%20%7B%0A%20%20%20%20bar%3A%20T%2C%0A%7D%0A%0Aimpl%3CT%3E%20Foo%3CT%3E%20%7B%0A%20%20%20%20fn%20barify%3CT%3E(%26self%2C%20bar%3A%20T)%20%7B%0A%20%20%20%20%20%20%20%20unimplemented!()%3B%0A%20%20%20%20%7D%0A%7D&version=stable&backtrace=0) says

> type parameter `T` shadows another type parameter of the same name